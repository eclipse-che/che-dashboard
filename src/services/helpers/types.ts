/*
 * Copyright (c) 2018-2020 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { AlertVariant } from '@patternfly/react-core';
import * as React from 'react';

export interface AlertItem {
  key: string;
  title: string;
  variant: AlertVariant;
  children?: React.ReactNode;
}

export interface FactoryResolver {
  v: string;
  source: string;
  devfile: api.che.workspace.devfile.Devfile
}

export enum WorkspaceStatus {
  RUNNING = 1,
  STOPPING,
  STOPPED,
  STARTING,
  PAUSED,
  ERROR,
}

export enum IdeLoaderTabs {
  Progress = 0,
  Logs = 1,
}
