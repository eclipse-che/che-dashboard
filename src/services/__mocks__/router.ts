/*
 * Copyright (c) 2018-2020 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { createLocation, createMemoryHistory, Location, LocationState } from 'history';
import { MemoryHistory } from 'history/createMemoryHistory';
import { match as routerMatch } from 'react-router';

const generateUrl = <Params extends { [K in keyof Params]?: string } = {}>(path: string, params: Params): string => {
  let resultPath = path;
  for (const param in params) {
    if (params[param]) {
      resultPath = resultPath.replace(`:${param}`, `${params[param]}`);
    }
  }
  return resultPath;
};

export const getMockRouterProps = <Params extends { [K in keyof Params]: string } = {}>(path: string, params: Params): {
  history: MemoryHistory<LocationState>;
  location: Location<LocationState>;
  match: routerMatch<Params>
} => {
  const isExact = false;
  const url = generateUrl(path, params);

  const match: routerMatch<Params> = { isExact, path, url, params };
  const history = createMemoryHistory();
  const location = createLocation(match.url);

  return { history, location, match };
};
