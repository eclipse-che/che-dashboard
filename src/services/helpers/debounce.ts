/*
 * Copyright (c) 2018-2020 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { injectable } from 'inversify';

/**
 * This class is handling the debounce delay service.
 * @author Oleksii Orel
 */
@injectable()
export class Debounce {
  private debounceTimer: any;
  private isDebounceDelay = false;
  private debounceDelayHandlers: Array<Function> = [];

  setDelay(timeDelay = 5000): void {
    this.setDebounceDelay(true);
    if (this.debounceTimer) {
      clearTimeout(this.debounceTimer);
    }
    this.debounceTimer = setTimeout(() => {
      this.setDebounceDelay(false);
    }, timeDelay);
  }

  private setDebounceDelay(isDebounceDelay: boolean): void {
    this.isDebounceDelay = isDebounceDelay;
    this.debounceDelayHandlers.forEach(handler => {
      if (typeof handler === 'function') {
        handler(isDebounceDelay);
      }
    });
  }

  /**
   * Subscribe on the debounce delay event.
   * @param handler
   */
  subscribe(handler: (isDebounceDelay: boolean) => void): void {
    this.debounceDelayHandlers.push(handler);
  }

  /**
   * Unsubscribe on the show alert event.
   * @param handler
   */
  public unsubscribe(handler: (isDebounceDelay: boolean) => void): void {
    const index = this.debounceDelayHandlers.indexOf(handler);
    if (index !== -1) {
      this.debounceDelayHandlers.splice(index, 1);
    }
  }

  /**
   * Unsubscribe all.
   */
  unsubscribeAll(): void {
    this.debounceDelayHandlers = [];
  }

  hasDelay(): boolean {
    return this.isDebounceDelay;
  }
}
