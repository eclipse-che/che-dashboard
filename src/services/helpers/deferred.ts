/*
 * Copyright (c) 2018-2020 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

export type IResolveFn<T> = {
  (value?: T | PromiseLike<T>): void;
}

export type IRejectFn<T> = {
  (reason?: any): void;
}

/**
 * Provides a Defer object.
 */
export type IDeferred<T> = {
  resolve(value?: T | Promise<T>): void;
  reject(reason?: any): void;
  promise: Promise<T>;
}

export const getDefer = <T>(): IDeferred<T> => {
  const defer: { [param: string]: any } = {};
  defer.promise = new Promise((resolve: IResolveFn<T>, reject: IRejectFn<any>) => {
    defer.resolve = resolve;
    defer.reject = reject;
  });
  return defer as IDeferred<T>;
};
