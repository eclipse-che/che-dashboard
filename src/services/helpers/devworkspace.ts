/*
 * Copyright (c) 2018-2020 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { devWorkspaceToDevfile, IDevWorkspace, IDevWorkspaceDevfile } from '@eclipse-che/devworkspace-client';
import { DevWorkspaceStatus, WorkspaceStatus } from './types';

/**
 * Convert a devworkspace to something that the natively dashboard understands
 * @param devworkspace The devworkspace that you want to convert
 */
export function convertDevWorkspaceV2ToV1(devworkspace: IDevWorkspace): che.Workspace {
  const convertedWorkspace = {} as che.Workspace;
  const namespace = devworkspace.metadata.namespace;
  convertedWorkspace.namespace = namespace;
  convertedWorkspace.devfile = devWorkspaceToDevfile(devworkspace);
  const epochCreatedTimestamp = new Date(devworkspace.metadata.creationTimestamp as string).valueOf();
  convertedWorkspace.attributes = {
    infrastructureNamespace: namespace,
    created: epochCreatedTimestamp.toString(),
  };
  if (devworkspace.status?.devworkspaceId) {
    convertedWorkspace.id = devworkspace.status?.devworkspaceId;
  }
  let status = devworkspace.status?.phase;
  if (status) {
    status = status.toUpperCase();
    if (status === DevWorkspaceStatus[DevWorkspaceStatus.FAILED]) {
      status = WorkspaceStatus[WorkspaceStatus.ERROR];
    }
    convertedWorkspace.status = status;
  }

  if (status && devworkspace.status?.ideUrl) {
    convertedWorkspace.runtime = {
      status,
      activeEnv: '',
      machines: {
        theia: {
          servers: {
            theia: {
              attributes: {
                type: 'ide'
              },
              url: devworkspace.status.ideUrl,
              status
            },
          },
          attributes: {},
          status
        }
      }
    };
  }
  return convertedWorkspace;
}

/**
 * Check to see if the workspace or devfile is a DevWorkspace
 * @param devworkspaceCustomResourceOrDevfile The devworkspace or devfile you want to check
 */
export function isDevWorkspace(devworkspaceCustomResourceOrDevfile: che.Workspace | api.che.workspace.devfile.Devfile | IDevWorkspaceDevfile): boolean {
  return (devworkspaceCustomResourceOrDevfile as any).kind === 'DevWorkspace' || (devworkspaceCustomResourceOrDevfile as any).devfile?.schemaVersion !== undefined || (devworkspaceCustomResourceOrDevfile as any).schemaVersion !== undefined;
}

/**
 * Check to see if the workspace is currently being deleted
 * @param workspace The workspace you want to check
 */
export function isDeleting(workspace: IDevWorkspace): boolean {
  return workspace.metadata.deletionTimestamp !== undefined;
}

/**
 * Check to see if the workspace is a web terminal
 * @param workspaceOrDevfile The workspace or devfile you want to check
 */
export function isWebTerminal(workspaceOrDevfile: che.Workspace | api.che.workspace.devfile.Devfile): boolean {
  return !!(workspaceOrDevfile as any).metadata.labels['console.openshift.io/terminal'];
}
