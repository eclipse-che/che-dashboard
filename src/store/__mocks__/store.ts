/*
 * Copyright (c) 2018-2020 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Store } from 'redux';
import createMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { AppState } from '../../store';

/**
 * @deprecated
 */
export const createFakeStore = (
  workspaces: che.Workspace[],
  workspacesLogs: Map<string, string[]> = new Map<string, string[]>()): Store => {
  const middleware = [thunk];
  const mockStore = createMockStore(middleware);
  return mockStore({
    factoryResolver: {
      isLoading: false,
      resolver: {},
    },
    plugins: {
      isLoading: false,
      plugins: [],
    },
    workspaces: {
      isLoading: false,
      settings: {} as any,
      workspaces,
      workspacesLogs,
      namespace: '',
      workspaceName: '',
      workspaceId: '',
      recentNumber: 5,
    },
    branding: {
      data: {
        docs: {
          devfile: 'https://www.test.docs/che'
        }
      }
    } as any,
    devfileRegistries: {} as any,
    user: {} as any,
    infrastructureNamespace: {} as any,
    userPreferences: {} as any,
  } as AppState);
};
