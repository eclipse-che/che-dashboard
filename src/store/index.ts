/*
 * Copyright (c) 2018-2020 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Action } from 'redux';
import { ThunkAction } from 'redux-thunk';
import * as Plugins from './Plugins';
import * as FactoryResolver from './FactoryResolver';
import * as WorkspacesStore from './Workspaces';
import * as DevfileRegistriesStore from './DevfileRegistries';
import * as InfrastructureNamespaceStore from './InfrastructureNamespace';
import * as BrandingStore from './Branding';
import * as UserPreferences from './UserPreferences';
import userReducer, { UserState } from './User';

// the top-level state object
export interface AppState {
  branding: BrandingStore.State;
  devfileRegistries: DevfileRegistriesStore.State;
  infrastructureNamespace: InfrastructureNamespaceStore.State;
  user: UserState;
  workspaces: WorkspacesStore.State;
  plugins: Plugins.State;
  factoryResolver: FactoryResolver.State;
  userPreferences: UserPreferences.State;
}

export const reducers = {
  workspaces: WorkspacesStore.reducer,
  devfileRegistries: DevfileRegistriesStore.reducer,
  branding: BrandingStore.reducer,
  user: userReducer,
  infrastructureNamespace: InfrastructureNamespaceStore.reducer,
  plugins: Plugins.reducer,
  factoryResolver: FactoryResolver.reducer,
  userPreferences: UserPreferences.reducer,
};

// this type can be used as a hint on action creators so that its 'dispatch' and 'getState' params are
// correctly typed to match your store.
/**
 * @deprecated
 */
export interface AppThunkAction<TAction> {
  (dispatch: (action: TAction) => void, getState: () => AppState): void;
}
export type AppThunk<ActionType extends Action, ReturnType = void> = ThunkAction<
  ReturnType,
  AppState,
  unknown,
  ActionType
>;
