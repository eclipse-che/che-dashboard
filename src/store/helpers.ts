/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { FactoryResolver } from '../services/helpers/types';
import { isDevfileV2 } from '../services/workspaceAdapter';

const PROJECT_NAME_MAX_LENGTH = 63;

type ProjectV2Source = {
  name: string;
  git: {
    remotes: { origin: string };
    checkoutFrom?: { revision: string };
  }
};

/**
 * Creates a new state object.
 * @param state a store state, e.g workspaces, plugins.
 * @param partial a slice of a store state
 */
export function createState<T>(state: T, partial: Partial<T>): T {
  return Object.assign({}, state, partial);
}

/**
 * Returns a devfile from the FactoryResolver object.
 * @param data a FactoryResolver object.
 */
export function getDevfile(data: FactoryResolver): api.che.workspace.devfile.Devfile {
  let devfile = data.devfile;

  if (isDevfileV2(devfile)) {
    devfile = Object.assign(devfile, { components });
    // add a default project
    const projects: ProjectV2Source[] = [];
    const scmInfo = data['scm_info'];
    if (!devfile.projects?.length && scmInfo) {
      const origin = scmInfo['clone_url'];
      const name = getProjectName(origin);
      const revision = scmInfo.branch;
      const project: ProjectV2Source = { name, git: { remotes: { origin } } };
      if (revision) {
        project.git.checkoutFrom = { revision };
      }
      projects.push(project);
      devfile = Object.assign({ projects }, devfile);
    }
  }

  return devfile;
}

export function getProjectName(cloneUrl: string): string {
  let name = cloneUrl.split('/').reverse()[0].replace(/(.git)$/, '');
  name = name.replace(/([^-a-zA-Z0-9]+)/, '-');
  name = name.replace(/(^[-]+)/, '');
  name = name.replace(/([-]+$)/, '');
  if (name.length > PROJECT_NAME_MAX_LENGTH) {
    name = name.substr(0, PROJECT_NAME_MAX_LENGTH - 1);
  }

  return name;
}
