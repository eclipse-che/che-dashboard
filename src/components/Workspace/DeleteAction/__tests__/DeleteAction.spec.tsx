/*
 * Copyright (c) 2018-2020 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { Provider } from 'react-redux';
import renderer, { ReactTestRendererJSON } from 'react-test-renderer';
import { render, RenderResult, screen } from '@testing-library/react';
import { WorkspaceDeleteAction } from '../';
import { WorkspaceStatus } from '../../../../services/helpers/types';
import { createFakeStore } from '../../../../store/__mocks__/store';

describe('Workspace delete component', () => {
  const workspaceId = 'workspace-test-id';

  it('should render delete widget enabled correctly', () => {
    const status = WorkspaceStatus.STOPPED;
    const disabled = false;
    const component = createComponent(status, disabled, workspaceId, jest.fn(), jest.fn());

    expect(getComponentSnapshot(component)).toMatchSnapshot();
  });

  it('should render delete widget disabled correctly', () => {
    const status = WorkspaceStatus.STOPPED;
    const disabled = false;
    const component = createComponent(status, disabled, workspaceId, jest.fn(), jest.fn());

    expect(getComponentSnapshot(component)).toMatchSnapshot();
  });

  it('should delete workspace on start correctly', () => {
    const status = WorkspaceStatus.STOPPED;
    const disabled = false;
    WorkspaceDeleteAction.shouldDelete.push(workspaceId);
    const deleteWorkspace = jest.fn();
    const stopWorkspace = jest.fn();

    renderComponent(createComponent(status, disabled, workspaceId, deleteWorkspace, stopWorkspace));

    expect(stopWorkspace).not.toBeCalled();
    expect(deleteWorkspace).toBeCalledWith(workspaceId);
  });

  it('should delete workspace if enable', () => {
    const status = WorkspaceStatus.STOPPED;
    const disabled = false;
    const deleteWorkspace = jest.fn();
    const stopWorkspace = jest.fn();

    renderComponent(createComponent(status, disabled, workspaceId, deleteWorkspace, stopWorkspace));

    expect(stopWorkspace).not.toBeCalled();
    expect(deleteWorkspace).not.toBeCalled();

    const deleteWorkspaceButton = screen.getByTestId(`delete-${workspaceId}`);
    deleteWorkspaceButton.click();

    const warningInfoCheckbox = screen.getByTestId('warning-info-checkbox');
    warningInfoCheckbox.click();

    const deleteWarningInfoButton = screen.getByTestId('delete-button');
    deleteWarningInfoButton.click();

    expect(stopWorkspace).not.toBeCalled();
    expect(deleteWorkspace).toBeCalledWith(workspaceId);
  });

  it('shouldn\'t delete workspace if disabled', () => {
    const status = WorkspaceStatus.STOPPED;
    const disabled = true;
    const deleteWorkspace = jest.fn();
    const stopWorkspace = jest.fn();

    renderComponent(createComponent(status, disabled, workspaceId, deleteWorkspace, stopWorkspace));

    expect(stopWorkspace).not.toBeCalled();
    expect(deleteWorkspace).not.toBeCalled();

    const deleteWorkspaceButton = screen.getByTestId(`delete-${workspaceId}`);
    deleteWorkspaceButton.click();

    expect(stopWorkspace).not.toBeCalled();
    expect(deleteWorkspace).not.toBeCalled();
  });

  it('shouldn\'t delete workspace without warning-info checkbox', () => {
    WorkspaceDeleteAction.shouldDelete.length = 0;
    const deleteWorkspace = jest.fn();
    const stopWorkspace = jest.fn();

    const status = WorkspaceStatus.RUNNING;

    renderComponent(createComponent(status, false, workspaceId, deleteWorkspace, stopWorkspace));

    expect(stopWorkspace).not.toBeCalled();
    expect(deleteWorkspace).not.toBeCalled();
    expect(WorkspaceDeleteAction.shouldDelete).toEqual([]);

    const getStartedTabButton = screen.getByTestId(`delete-${workspaceId}`);
    getStartedTabButton.click();

    const deleteWarningInfoButton = screen.getByTestId('delete-button');
    deleteWarningInfoButton.click();

    expect(deleteWorkspace).not.toBeCalled();
    expect(stopWorkspace).not.toBeCalled();
    expect(WorkspaceDeleteAction.shouldDelete).toEqual([]);
  });

  it('should stop workspace then delete if enable', () => {
    WorkspaceDeleteAction.shouldDelete.length = 0;
    let deleteWorkspace = jest.fn();
    let stopWorkspace = jest.fn();

    let status = WorkspaceStatus.RUNNING;

    const component = renderComponent(createComponent(status, false, workspaceId, deleteWorkspace, stopWorkspace));

    expect(stopWorkspace).not.toBeCalled();
    expect(deleteWorkspace).not.toBeCalled();
    expect(WorkspaceDeleteAction.shouldDelete).toEqual([]);

    const getStartedTabButton = screen.getByTestId(`delete-${workspaceId}`);
    getStartedTabButton.click();

    const warningInfoCheckbox = screen.getByTestId('warning-info-checkbox');
    warningInfoCheckbox.click();

    const deleteWarningInfoButton = screen.getByTestId('delete-button');
    deleteWarningInfoButton.click();

    expect(deleteWorkspace).not.toBeCalled();
    expect(stopWorkspace).toBeCalledWith(workspaceId);
    expect(WorkspaceDeleteAction.shouldDelete).toEqual([workspaceId]);

    status = WorkspaceStatus.STOPPED;
    deleteWorkspace = jest.fn();
    stopWorkspace = jest.fn();

    component.rerender(createComponent(status, false, workspaceId, deleteWorkspace, stopWorkspace));

    expect(stopWorkspace).not.toBeCalled();
    expect(deleteWorkspace).toBeCalledWith(workspaceId);
  });

});

function createComponent(
  workspaceStatus: WorkspaceStatus,
  disabled: boolean,
  workspaceId: string,
  deleteWorkspace: jest.Mock,
  stopWorkspace: jest.Mock,
): React.ReactElement {
  const store = createFakeStore([]);
  return (
    <Provider store={store}>
      <WorkspaceDeleteAction
        status={workspaceStatus}
        disabled={disabled}
        workspaceId={workspaceId}
        requestWorkspaces={jest.fn()}
        requestWorkspace={jest.fn()}
        startWorkspace={jest.fn()}
        stopWorkspace={async (id: string) => stopWorkspace(id)}
        deleteWorkspace={async (id: string) => deleteWorkspace(id)}
        updateWorkspace={jest.fn()}
        createWorkspaceFromDevfile={jest.fn()}
        requestSettings={jest.fn()}
        setWorkspaceQualifiedName={jest.fn()}
        clearWorkspaceQualifiedName={jest.fn()}
        setWorkspaceId={jest.fn()}
        clearWorkspaceId={jest.fn()}
        deleteWorkspaceLogs={jest.fn()}
      >Delete Workspace</WorkspaceDeleteAction>
    </Provider>
  );
}

function renderComponent(
  component: React.ReactElement
): RenderResult {
  return render(component);
}

function getComponentSnapshot(
  component: React.ReactElement
): null | ReactTestRendererJSON | ReactTestRendererJSON[] {
  return renderer.create(component).toJSON();
}
