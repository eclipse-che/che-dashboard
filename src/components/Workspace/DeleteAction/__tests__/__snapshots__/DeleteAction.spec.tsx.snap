// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workspace delete component should render delete widget disabled correctly 1`] = `
<span
  data-testid="delete-workspace-test-id"
  onClick={[Function]}
>
  Delete Workspace
</span>
`;

exports[`Workspace delete component should render delete widget enabled correctly 1`] = `
<span
  data-testid="delete-workspace-test-id"
  onClick={[Function]}
>
  Delete Workspace
</span>
`;
